### ISSUE TITLE
Presto field sync skips remaining tables when encountering a Hive view (or any other table field error)

### ISSUE THREAD

**Issue by lukeman** [metabase/metabase#11422]
**Describe the bug**
With a Presto connection, if the field sync encounters an exception with any of the discovered tables from the table sync step (such as when the "table" is actually a Hive view), it skips the remaining tables and marks the sync as complete.

**Logs**
```
[5b472577-7f2a-405e-9f2b-4b71fe9715f4] 2019-12-02T03:02:30-05:00 INFO metabase.sync.util FINISHED: step 'sync-tables' for presto Database 6 'GCP Presto' (1.9 mins)
[5b472577-7f2a-405e-9f2b-4b71fe9715f4] 2019-12-02T03:02:30-05:00 INFO metabase.sync.util STARTING: step 'sync-fields' for presto Database 6 'GCP Presto'
[5b472577-7f2a-405e-9f2b-4b71fe9715f4] 2019-12-02T03:03:35-05:00 ERROR metabase.sync.util Error syncing Fields for Table 'Table 129,440 'foo.bar'': clojure.lang.ExceptionInfo: Hive views are not supported: 'foo.bar' {:message "Hive views are not supported: 'foo.bar'", :errorCode 5, :errorName "NOT_FOUND", :errorType "USER_ERROR", :failureInfo {:type "com.facebook.presto.hive.HiveViewNotSupportedException", :message "Hive views are not supported: 'foo.bar'", :suppressed [], :stack ["com.facebook.presto.hive.metastore.thrift.ThriftHiveMetastore.lambda$getTable$5(ThriftHiveMetastore.java:223)" "com.facebook.presto.hive.metastore.thrift.HiveMetastoreApiStats.lambda$wrap$0(HiveMetastoreApiStats.java:42)" "com.facebook.presto.hive.RetryDriver.run(RetryDriver.java:137)" "com.facebook.presto.hive.metastore.thrift.ThriftHiveMetastore.getTable(ThriftHiveMetastore.java:219)" "com.facebook.presto.hive.metastore.thrift.BridgingHiveMetastore.getTable(BridgingHiveMetastore.java:78)" "com.facebook.presto.hive.metastore.CachingHiveMetastore.loadTable(CachingHiveMetastore.java:268)" "com.google.common.cache.CacheLoader$FunctionToCacheLoader.load(CacheLoader.java:165)" "com.google.common.cache.CacheLoader$1.load(CacheLoader.java:188)" "com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3524)" "com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2273)" "com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2156)" "com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2046)" "com.google.common.cache.LocalCache.get(LocalCache.java:3943)" "com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3967)" "com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4952)" "com.google.common.cache.LocalCache$LocalLoadingCache.getUnchecked(LocalCache.java:4958)" "com.facebook.presto.hive.metastore.CachingHiveMetastore.get(CachingHiveMetastore.java:212)" "com.facebook.presto.hive.metastore.CachingHiveMetastore.getTable(CachingHiveMetastore.java:257)" "com.facebook.presto.hive.metastore.CachingHiveMetastore.loadTable(CachingHiveMetastore.java:268)" "com.google.common.cache.CacheLoader$FunctionToCacheLoader.load(CacheLoader.java:165)" "com.google.common.cache.CacheLoader$1.load(CacheLoader.java:188)" "com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3524)" "com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2273)" "com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2156)" "com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2046)" "com.google.common.cache.LocalCache.get(LocalCache.java:3943)" "com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3967)" "com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4952)" "com.google.common.cache.LocalCache$LocalLoadingCache.getUnchecked(LocalCache.java:4958)" "com.facebook.presto.hive.metastore.CachingHiveMetastore.get(CachingHiveMetastore.java:212)" "com.facebook.presto.hive.metastore.CachingHiveMetastore.getTable(CachingHiveMetastore.java:257)" "com.facebook.presto.hive.metastore.SemiTransactionalHiveMetastore.getTable(SemiTransactionalHiveMetastore.java:139)" "com.facebook.presto.hive.HiveMetadata.getViews(HiveMetadata.java:1281)" "com.facebook.presto.spi.connector.classloader.ClassLoaderSafeConnectorMetadata.getViews(ClassLoaderSafeConnectorMetadata.java:337)" "com.facebook.presto.metadata.MetadataManager.getView(MetadataManager.java:808)" "com.facebook.presto.sql.rewrite.ShowQueriesRewrite$Visitor.visitShowColumns(ShowQueriesRewrite.java:304)" "com.facebook.presto.sql.rewrite.ShowQueriesRewrite$Visitor.visitShowColumns(ShowQueriesRewrite.java:141)" "com.facebook.presto.sql.tree.ShowColumns.accept(ShowColumns.java:54)" "com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:27)" "com.facebook.presto.sql.rewrite.ShowQueriesRewrite.rewrite(ShowQueriesRewrite.java:138)" "com.facebook.presto.sql.rewrite.StatementRewrite.rewrite(StatementRewrite.java:51)" "com.facebook.presto.sql.analyzer.Analyzer.analyze(Analyzer.java:69)" "com.facebook.presto.sql.analyzer.Analyzer.analyze(Analyzer.java:64)" "com.facebook.presto.execution.SqlQueryExecution.<init>(SqlQueryExecution.java:169)" "com.facebook.presto.execution.SqlQueryExecution$SqlQueryExecutionFactory.createQueryExecution(SqlQueryExecution.java:677)" "com.facebook.presto.execution.SqlQueryManager.createQueryInternal(SqlQueryManager.java:444)" "com.facebook.presto.execution.SqlQueryManager.lambda$createQuery$3(SqlQueryManager.java:377)" "java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)" "java.util.concurrent.FutureTask.run(FutureTask.java:266)" "java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)" "java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)" "java.lang.Thread.run(Thread.java:748)"]}}
("java.util.concurrent.FutureTask.report(FutureTask.java:122)"
"java.util.concurrent.FutureTask.get(FutureTask.java:192)"
"clojure.core$deref_future.invokeStatic(core.clj:2300)"
"clojure.core$future_call$reify__8454.deref(core.clj:6974)"
"clojure.core$deref.invokeStatic(core.clj:2320)"
"clojure.core$deref.invoke(core.clj:2306)"
"--> driver.presto$execute_presto_query_BANG_$fn__1083.invoke(presto.clj:141)"
"util.ssh$with_ssh_tunnel_STAR_.invokeStatic(ssh.clj:106)"
"util.ssh$with_ssh_tunnel_STAR_.invoke(ssh.clj:91)"
"driver.presto$execute_presto_query_BANG_.invokeStatic(presto.clj:125)"
"driver.presto$execute_presto_query_BANG_.invoke(presto.clj:121)"
"driver.presto$eval1213$fn__1216.invoke(presto.clj:220)"
"sync.fetch_metadata$fn__49533$table_metadata__49538$fn__49539.invoke(fetch_metadata.clj:18)"
"sync.fetch_metadata$fn__49533$table_metadata__49538.invoke(fetch_metadata.clj:15)"
"sync.sync_metadata.fields.fetch_metadata$fn__49855$db_metadata__49860$fn__49861.invoke(fetch_metadata.clj:82)"
"sync.sync_metadata.fields.fetch_metadata$fn__49855$db_metadata__49860.invoke(fetch_metadata.clj:78)"
"sync.sync_metadata.fields$fn__50359$sync_fields_for_table_BANG___50368$fn__50374$fn__50376.invoke(fields.clj:91)"
"sync.util$do_with_error_handling.invokeStatic(util.clj:149)"
"sync.util$do_with_error_handling.invoke(util.clj:144)"
"sync.sync_metadata.fields$fn__50359$sync_fields_for_table_BANG___50368$fn__50374.invoke(fields.clj:90)"
"sync.sync_metadata.fields$fn__50359$sync_fields_for_table_BANG___50368.invoke(fields.clj:83)"
"sync.sync_metadata.fields$fn__50404$sync_fields_BANG___50409$fn__50410$iter__50411__50415$fn__50416$fn__50417.invoke(fields.clj:117)"
"sync.sync_metadata.fields$fn__50404$sync_fields_BANG___50409$fn__50410$iter__50411__50415$fn__50416.invoke(fields.clj:116)"
"sync.sync_metadata.fields$fn__50404$sync_fields_BANG___50409$fn__50410.invoke(fields.clj:116)"
"sync.sync_metadata.fields$fn__50404$sync_fields_BANG___50409.invoke(fields.clj:111)"
"sync.util$fn__41410$run_step_with_metadata__41415$fn__41419$fn__41421.invoke(util.clj:356)"
"sync.util$with_start_and_finish_logging_STAR_.invokeStatic(util.clj:104)"
"sync.util$with_start_and_finish_logging_STAR_.invoke(util.clj:98)"
"sync.util$with_start_and_finish_debug_logging.invokeStatic(util.clj:121)"
"sync.util$with_start_and_finish_debug_logging.invoke(util.clj:118)"
"sync.util$fn__41410$run_step_with_metadata__41415$fn__41419.invoke(util.clj:353)"
"sync.util$fn__41410$run_step_with_metadata__41415.invoke(util.clj:348)"
"sync.util$fn__41602$run_sync_operation__41607$fn__41608$fn__41609.invoke(util.clj:441)"
"sync.util$fn__41602$run_sync_operation__41607$fn__41608.invoke(util.clj:441)"
"sync.util$fn__41602$run_sync_operation__41607.invoke(util.clj:435)"
"sync.sync_metadata$fn__51091$sync_db_metadata_BANG___51096$fn__51097$fn__51098.invoke(sync_metadata.clj:51)"
"sync.util$do_with_error_handling.invokeStatic(util.clj:149)"
"sync.util$do_with_error_handling.invoke(util.clj:144)"
"sync.util$do_with_error_handling.invokeStatic(util.clj:147)"
"sync.util$do_with_error_handling.invoke(util.clj:144)"
"driver$fn__19284.invokeStatic(driver.clj:673)"
"driver$fn__19284.invoke(driver.clj:673)"
"sync.util$sync_in_context$fn__41316.invoke(util.clj:140)"
"sync.util$with_db_logging_disabled$fn__41313.invoke(util.clj:131)"
"sync.util$with_start_and_finish_logging_STAR_.invokeStatic(util.clj:104)"
"sync.util$with_start_and_finish_logging_STAR_.invoke(util.clj:98)"
"sync.util$with_start_and_finish_logging$fn__41302.invoke(util.clj:116)"
"sync.util$with_sync_events$fn__41297.invoke(util.clj:90)"
"sync.util$with_duplicate_ops_prevented$fn__41288.invoke(util.clj:69)"
"sync.util$do_sync_operation.invokeStatic(util.clj:168)"
"sync.util$do_sync_operation.invoke(util.clj:165)"
"sync.sync_metadata$fn__51091$sync_db_metadata_BANG___51096$fn__51097.invoke(sync_metadata.clj:50)"
"sync.sync_metadata$fn__51091$sync_db_metadata_BANG___51096.invoke(sync_metadata.clj:47)"
"task.sync_databases.SyncAndAnalyzeDatabase.execute(sync_databases.clj:39)")

[5b472577-7f2a-405e-9f2b-4b71fe9715f4] 2019-12-02T03:03:35-05:00 ERROR metabase.sync.util Error running sync step: Output of sync-fields-for-table! does not match schema:

(not (map? a-java.util.concurrent.ExecutionException))


("--> sync.sync_metadata.fields$fn__50359$sync_fields_for_table_BANG___50368.invoke(fields.clj:83)"
"sync.sync_metadata.fields$fn__50404$sync_fields_BANG___50409$fn__50410$iter__50411__50415$fn__50416$fn__50417.invoke(fields.clj:117)"
"sync.sync_metadata.fields$fn__50404$sync_fields_BANG___50409$fn__50410$iter__50411__50415$fn__50416.invoke(fields.clj:116)"
"sync.sync_metadata.fields$fn__50404$sync_fields_BANG___50409$fn__50410.invoke(fields.clj:116)"
"sync.sync_metadata.fields$fn__50404$sync_fields_BANG___50409.invoke(fields.clj:111)"
"sync.util$fn__41410$run_step_with_metadata__41415$fn__41419$fn__41421.invoke(util.clj:356)"
"sync.util$with_start_and_finish_logging_STAR_.invokeStatic(util.clj:104)"
"sync.util$with_start_and_finish_logging_STAR_.invoke(util.clj:98)"
"sync.util$with_start_and_finish_debug_logging.invokeStatic(util.clj:121)"
"sync.util$with_start_and_finish_debug_logging.invoke(util.clj:118)"
"sync.util$fn__41410$run_step_with_metadata__41415$fn__41419.invoke(util.clj:353)"
"sync.util$fn__41410$run_step_with_metadata__41415.invoke(util.clj:348)"
"sync.util$fn__41602$run_sync_operation__41607$fn__41608$fn__41609.invoke(util.clj:441)"
"sync.util$fn__41602$run_sync_operation__41607$fn__41608.invoke(util.clj:441)"
"sync.util$fn__41602$run_sync_operation__41607.invoke(util.clj:435)"
"sync.sync_metadata$fn__51091$sync_db_metadata_BANG___51096$fn__51097$fn__51098.invoke(sync_metadata.clj:51)"
"sync.util$do_with_error_handling.invokeStatic(util.clj:149)"
"sync.util$do_with_error_handling.invoke(util.clj:144)"
"sync.util$do_with_error_handling.invokeStatic(util.clj:147)"
"sync.util$do_with_error_handling.invoke(util.clj:144)"
"driver$fn__19284.invokeStatic(driver.clj:673)"
"driver$fn__19284.invoke(driver.clj:673)"
"sync.util$sync_in_context$fn__41316.invoke(util.clj:140)"
"sync.util$with_db_logging_disabled$fn__41313.invoke(util.clj:131)"
"sync.util$with_start_and_finish_logging_STAR_.invokeStatic(util.clj:104)"
"sync.util$with_start_and_finish_logging_STAR_.invoke(util.clj:98)"
"sync.util$with_start_and_finish_logging$fn__41302.invoke(util.clj:116)"
"sync.util$with_sync_events$fn__41297.invoke(util.clj:90)"
"sync.util$with_duplicate_ops_prevented$fn__41288.invoke(util.clj:69)"
"sync.util$do_sync_operation.invokeStatic(util.clj:168)"
"sync.util$do_sync_operation.invoke(util.clj:165)"
"sync.sync_metadata$fn__51091$sync_db_metadata_BANG___51096$fn__51097.invoke(sync_metadata.clj:50)"
"sync.sync_metadata$fn__51091$sync_db_metadata_BANG___51096.invoke(sync_metadata.clj:47)"
"task.sync_databases.SyncAndAnalyzeDatabase.execute(sync_databases.clj:39)")

[5b472577-7f2a-405e-9f2b-4b71fe9715f4] 2019-12-02T03:03:35-05:00 INFO metabase.sync.util FINISHED: Sync metadata for presto Database 6 'GCP Presto' (3.0 mins)
```

**To Reproduce**
1. Add a Presto database with at least one Hive view in the metastore
2. Kick off a schema sync
3. Observe sync errors in the logs and try browse any table that landed after the first Hive view in the table discovery sync step

**Expected behavior**
When the field sync triggers an exception on an individual table (which seems to be expected behavior with some Presto connectors), I'd expect the process to continue on to the next table in the list.

**Information about your Metabase Installation:**

- Your databases: Presto
- Metabase version: 0.33.6
- Metabase hosting environment: Docker
- Metabase internal database: Postgres

**Severity**
Raw queries still work, but this renders the question editor and data browse/metadata portions non-functional for some amount of tables (most of them in my particular case).

**Comment by camsaul** [comment]
Generally if a sync substep fails we should continue rather than failing sync entirely. e.g. if syncing one Table fails we should still sync the rest. So for this it seems like we just need to come up with a simple test case to repro this and then fix the sync error handling logic.

**Comment by sbelak** [comment]
I think this hardening was already done in #11820

**Comment by robdaemon** [comment]
I believe this was fixed - I've tried reproducing it against 0.36.2 and everything seems to be working. It looks like we fixed this in 0.35.0.

I've set up a local Hive and PrestoDB instance using [this Docker setup](https://github.com/big-data-europe/docker-hive), and connected them to Metabase with the following schemas:

```sql
CREATE TABLE pokes (foo INT, bar STRING);
LOAD DATA LOCAL INPATH '/opt/hive/examples/files/kv1.txt' OVERWRITE INTO TABLE pokes;

CREATE VIEW pview (foo, bar) AS SELECT * FROM pokes;

CREATE TABLE zpoke (foo INT, bar STRING);
LOAD DATA LOCAL INPATH '/opt/hive/examples/files/kv1.txt' OVERWRITE INTO TABLE zpoke;
```

The error about Hive views not being supported is present, but this is a limitation of the Presto connector.

```text
08-07 19:11:33 ERROR sync.util :: Error syncing Fields for Table 'Table 18 'default.pview''
clojure.lang.ExceptionInfo: Hive views are not supported: 'default.pview' {:message "Hive views are not supported: 'default.pview'", :errorCode 5, :errorName "NOT_FOUND", :errorType "USER_ERROR", :failureInfo {:type "com.facebook.presto.hive.HiveViewNotSupportedException", :message "Hive views are not supported: 'default.pview'",
```

Metabase continues on after that error and syncs the rest of the database, and the remaining datasources.

I'm going to close this for now, but let us know if you still have a problem here. Thanks!

**Comment by lukeman** [comment]
Just wanted to confirm that things look great now (couldn't confirm before as I'd broken schema sync on my instance with a temporary rollback to v0.35). Thanks y'all.
